internal/cluster/distribution/eks/eksworkflow/activity_update_addon.go

```diff
@@ -143,33 +146,63 @@ func (a *UpdateAddonActivity) Execute(ctx context.Context, input UpdateAddonActi
 		}
 		return nil, errors.WrapIfWithDetails(err, "failed to update addon", "cluster", input.ClusterName, "addon", input.AddonName)
 	}
-	output := UpdateAddonActivityOutput{UpdateID: aws.StringValue(updateAddonOutput.Update.Id)}
+	output := UpdateAddonActivityOutput{UpdateID: aws.StringValue(updateAddonOutput.Update.Id), LatestVersion: latestVersion}
 
 	return &output, nil
 }
 
-func selectLatestVersion(addonVersions *eks.DescribeAddonVersionsOutput, currentVersion string, kubernetesVersion string) (string, error) {
+// return latest available version which might be the currentVersion. In case selectNextMinor = true tries to return next available minor version.
+// Beside the next version it returns whether the returned version is the latest one.
+func selectNextVersion(addonVersions *eks.DescribeAddonVersionsOutput, currentVersion string, kubernetesVersion string, selectNextMinor bool) (string, bool, error) {
 	currentVersionSemver, err := semver.NewVersion(currentVersion)
 	if err != nil {
-		return "", err
+		return "", false, err
 	}
 	latestVersion := currentVersionSemver
 
+	versions := make([]*semver.Version, 0, 5)
 	for _, addon := range addonVersions.Addons {
 		for _, version := range addon.AddonVersions {
 			if !versionIsCompatible(version.Compatibilities, kubernetesVersion) {
 				continue
 			}
-			newVersion, err := semver.NewVersion(*version.AddonVersion)
+			version, err := semver.NewVersion(*version.AddonVersion)
 			if err != nil {
-				return "", err
+				return "", false, err
 			}
-			if newVersion.GreaterThan(latestVersion) {
-				latestVersion = newVersion
+			if version.GreaterThan(latestVersion) {
+				versions = append(versions, version)
 			}
 		}
 	}
-	return latestVersion.Original(), nil
+
+	if len(versions) == 0 {
+		return latestVersion.Original(), true, nil
+	}
+
+	sort.Slice(versions, func(i, j int) bool {
+		if versions[i].Compare(versions[j]) == -1 {
+			return true
+		}
+		return false
+	})
+
+	if selectNextMinor {
+		for i, version := range versions {
+			if !isSameMinorVersion(version, latestVersion) {
+				return version.Original(), i+1 == len(versions), nil
+			}
+		}
+	}
+
+	return versions[len(versions)-1].Original(), true, nil
+}
```

Comment:
**Required**: we **MUST** adhere to the Golang comment [guidelines](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and [start the doc comments of entities with their name](https://golang.org/doc/effective_go#commentary).

**Recommended**: I think we **SHOULD** rename the parameter `selectNextMinor` and reword the function comment, because it is misleading the same way `UpgradeToNextMinorVersion` is, namely it falsely implies when set to true it will result in either no update or a minor version update which is not true, because we also allow patch only updates to happen in cases only those are available, so I suggest renaming it to `selectNextAtMostOneMinorVersion` and update the comment to reflect that.

**Recommended**: I think we **SHOULD** do this in a less complicated, more clean way, because while it is a series of simple steps, the number of serialized steps (and number of loops used) still make it less straightforward then a `selectLatestOrNextAtMostOneMinorVersion` operation should be and thus increases the cognitive complexity.

Suggestion regarding the comments above:
```golang
// selectNextVersion returns the next version the addon can be upgraded to based on the specified current and Kubernetes versions and possibly limiting the upgrade to at most a minor version when the provided selectNextAtMostMinorVersion is set to true or alternatively an error. 
//
// When no version is available, the function returns the current version with the latest compatible version indicator set to true.
func selectNextVersion(addonVersions *eks.DescribeAddonVersionsOutput, currentVersion string, kubernetesVersion string, selectNextAtMostOneMinorVersion bool) (string, bool, error) {
	currentVersionSemver, err := semver.NewVersion(currentVersion)
	if err != nil {
		return "", false, errors.WrapWithDetails(err, "invalid current version, not a semantic version", "currentVersion", currentVersion)
	}

	latestVersion := currentVersionSemver
    nextAtMostOneMinorVersion := currentVersionSemver

    for _, addon := range addonVersions.Addons {
        for _, version := range addon.AddonVersions {
            if version == nil ||
                !versionIsCompatible(version.Compatibilities, kubernetesVersion) {
                continue
            }

            addonVersion, err := semver.NewVersion(*version.AddonVersion)
            if err != nil {
                return "", false, errors.WrapWithDetails(err. "invalid addon version, not a semantic version", "addonVersion", *version.AddonVersion)
            }

        if addonVersion.Major() == currentVersionSemver.Major() && // Note: only consider same major version versions (multiple minor version updates at most).
            addonVersion.Minor() <= currentVersionSemver.Minor()+1 &&  // Note: only consider +1 minor version versions.
            addonVersion.GreaterThan(nextAtMostOneMinorVersion) { // Note: only set nextAtMostOneMinorVersion to higher version than it was before, initialized to currentVersionSemver.
            nextAtMostOneMinorVersion := addonVersion
        }

        if addonVersion.GreaterThan(latestVersion) {
            latestVersion := addonVersion
        }
    }
}

if selectNextAtMostOneMinorVersion {
    return nextAtMostOneMinorVersion.Original(), nextAtMostOneMinorVersion.Equal(latestVersion), nil
}

return latestVersion, true, nil
```
(GH suggestions are "not available for deleted lines", it says, that's why this is only a code block.)

**Note**: you **MUST** retest the functionality if you apply this change, because I wrote the code from heart based on the file content seen during the review.

Also `isSameMinorVersion()` is not required with this approach and can be removed.

Reply:
refactored